src/python/pants/source/source_root.py

```diff
@@ -474,3 +484,48 @@ def find(self, path) -> Optional[SourceRoot]:
                 return SourceRoot(path)
             # Otherwise, try the next value of i.
         return None
+
+
+@dataclass(frozen=True)
+class SourceRootRequest:
+    path: str
+
+    @classmethod
+    def for_file(cls, file_path: str) -> "SourceRootRequest":
```

Comment:
Nit: if you end up needing another CI run for this, a docstring for this method+class would be good.

Reply:
Done.